### PR TITLE
Fix bug in elastic net attack

### LINKED ISSUE #966: elastic net - low confidence adversarial examples
When performing elastic net attack, it seems that the parameter "confidence" is not used, so the attack does not return high confidence adversarial examples, even when one sets confidence to a high number.

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
Fix bug in elastic net attack.
Fixes #966 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/elastic_net.py [hunk 1]
```diff
@@ -144,16 +148,28 @@ def _gradient_of_loss(
         # Compute the current predictions
         predictions = self.estimator.predict(np.array(x_adv, dtype=ART_NUMPY_DTYPE), batch_size=self.batch_size)
 
+        # Then compute individual outputs
+        z_target = np.sum(predictions * target, axis=1)
+        z_other = np.max(
+            predictions * (1 - target) + (np.min(predictions, axis=1) - 1)[:, np.newaxis] * target,
+            axis=1,
+        )
+
```
Comment (beat-buesser) [inline art/attacks/evasion/elastic_net.py:157]:
```suggestion
```

File: art/attacks/evasion/elastic_net.py [hunk 2]
```diff
@@ -166,6 +182,9 @@ def _gradient_of_loss(
         loss_gradient *= c_mult
         loss_gradient += 2 * (x_adv - x)
 
+        # Set gradients where loss is constant to zero
+        loss_gradient[cond] = 0.0
```
Comment (beat-buesser) [inline art/attacks/evasion/elastic_net.py:186]:
```suggestion
        # Set gradients where loss is constant to zero
        cond = (predictions[i_add] - predictions[i_sub] + self.confidence) < 0
        loss_gradient[cond] = 0.0
```

File: art/attacks/evasion/elastic_net.py [hunk 3]
```diff
@@ -144,16 +148,28 @@ def _gradient_of_loss(
...
+            cond = (z_other - z_target + self.confidence) < 0
+
         else:
             i_add = np.argmax(target, axis=1)
             i_sub = np.argmax(
-                predictions * (1 - target) + (np.min(predictions, axis=1) - 1)[:, np.newaxis] * target, axis=1,
+                predictions * (1 - target) + (np.min(predictions, axis=1) - 1)[:, np.newaxis] * target,
+                axis=1,
             )
+            cond = (z_target - z_other + self.confidence) < 0
```
Comment (beat-buesser) [inline art/attacks/evasion/elastic_net.py:172]:
```suggestion
```

File: art/attacks/evasion/elastic_net.py [hunk 4]
```diff
@@ -144,16 +148,28 @@ def _gradient_of_loss(
...
+
         if self.targeted:
             i_sub = np.argmax(target, axis=1)
             i_add = np.argmax(
-                predictions * (1 - target) + (np.min(predictions, axis=1) - 1)[:, np.newaxis] * target, axis=1,
+                predictions * (1 - target) + (np.min(predictions, axis=1) - 1)[:, np.newaxis] * target,
+                axis=1,
             )
+            cond = (z_other - z_target + self.confidence) < 0
+
```
Comment (beat-buesser) [inline art/attacks/evasion/elastic_net.py:165]:
```suggestion
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @minhitbk Thank you very much for fixing this bug. I have a proposal to reuse the `i_add` and `i_sub` variables. What do you think, would this work?

**Comment by minhitbk** [comment]
> Hi @minhitbk Thank you very much for fixing this bug. I have a proposal to reuse the `i_add` and `i_sub` variables. What do you think, would this work?

Hi @beat-buesser : I don't think it works when you extract  predictions[i_sub] and predictions[i_add] as batch.

**Review by beat-buesser — APPROVED** [review]

**Comment by beat-buesser** [comment]
@minhitbk Thank you!

**Pull request merged**